research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py

```diff
@@ -0,0 +1,363 @@
+      channel_means = [123.68, 116.779, 103.939]
+      return resized_inputs - [[channel_means]]
+    else:
+      return resized_inputs
+
+  def get_proposal_feature_extractor_model(self, name=None):
+    """Returns a model that extracts first stage RPN features.
+
+    Extracts features using the Resnet v1 FPN network.
+
+    Args:
+      name: A scope name to construct all variables within.
+
+    Returns:
+      A Keras model that takes preprocessed_inputs:
+        A [batch, height, width, channels] float32 tensor
+        representing a batch of images.
+
+      And returns rpn_feature_map:
+        A list of tensors with shape [batch, height, width, depth]
+    """
+    with tf.name_scope(name):
+      with tf.name_scope('ResnetV1FPN'):
+        full_resnet_v1_model = self._resnet_v1_base_model(
+            batchnorm_training=self._train_batch_norm,
+            conv_hyperparams=(self._conv_hyperparams
+                              if self._override_base_feature_extractor_hyperparams
+                              else None),
+            min_depth=self._min_depth,
+            depth_multiplier=self._depth_multiplier,
+            classes=None,
+            weights=None,
+            include_top=False)
+        output_layers = _RESNET_MODEL_OUTPUT_LAYERS[self._resnet_v1_base_model_name]
+        outputs = [full_resnet_v1_model.get_layer(output_layer_name).output
+                   for output_layer_name in output_layers]
+        self.classification_backbone = tf.keras.Model(
+            inputs=full_resnet_v1_model.inputs,
+            outputs=outputs)
+        backbone_outputs = self.classification_backbone(full_resnet_v1_model.inputs)
+
+        # construct FPN feature generator
+        self._depth_fn = lambda d: max(
+            int(d * self._depth_multiplier), self._min_depth)
+        self._base_fpn_max_level = min(self._fpn_max_level, 5)
+        self._num_levels = self._base_fpn_max_level + 1 - self._fpn_min_level
+        self._fpn_features_generator = (
+            feature_map_generators.KerasFpnTopDownFeatureMaps(
+                num_levels=self._num_levels,
+                depth=self._depth_fn(self._additional_layer_depth),
+                is_training=self._is_training,
+                conv_hyperparams=self._conv_hyperparams,
+                freeze_batchnorm=self._freeze_batchnorm,
+                name='FeatureMaps'))
+
+        feature_block_list = []
+        for level in range(self._fpn_min_level, self._base_fpn_max_level + 1):
+          feature_block_list.append('block{}'.format(level - 1))
+        feature_block_map = dict(
+            list(zip(self._resnet_block_names, backbone_outputs)))
+        fpn_input_image_features = [
+            (feature_block, feature_block_map[feature_block])
+            for feature_block in feature_block_list]
+        fpn_features = self._fpn_features_generator(fpn_input_image_features)
+        features_maps = [fpn_feature for _, fpn_feature in fpn_features.items()]
+
```

Comment:
This is still missing coarse feature layers from self._base_fnp_max_level + 1 to self._fpn_max_level. Please see ssd_resnet_v1_fpn_keras_feature_extractor.py for reference.

Reply:
Thanks for adding!